File: code-generator/src/main/java/com/github/potjerodekool/openapi/LoggerFactory.java
package com.github.potjerodekool.openapi;

import java.util.function.Function;

public class LoggerFactory {

    private static final LoggerFactory FACTORY = new LoggerFactory();

    private Function<String, Logger> loggerProvider = DefaultLogger::new;

    private LoggerFactory() {
    }

    public static void setLoggerProvider(Function<String, Logger> loggerProvider) {
        FACTORY.loggerProvider = loggerProvider;
    }

    public static Logger getLogger(final String name) {
        return FACTORY.loggerProvider.apply(name);
    }

}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/log/LoggerFactory.kt
package io.github.potjerodekool.openapi.common.log

import java.util.function.Function

/**
 * A logger factory to create Logger instances.
 * Build systems should set the logger provider like this:
 * LoggerFactory#setLoggerProvider(mavenLoggerProvider)
 */
class LoggerFactory private constructor() {
    private var loggerProvider = Function { name: String -> devNullLoggerProvider(name) }

    companion object {
        private val FACTORY = LoggerFactory()

        fun setLoggerProvider(loggerProvider: Function<String, Logger>) {
            FACTORY.loggerProvider = loggerProvider
        }

        fun getLogger(name: String): Logger {
            return FACTORY.loggerProvider.apply(name)
        }

        private fun devNullLoggerProvider(name: String): Logger {
            return DevNullLogger()
        }
    }
}

class DevNullLogger : Logger {
    override fun log(level: LogLevel?, message: String?, exception: Throwable?) {
        //Nothing to do here
    }
}